src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/types.h"
+#include "model/record.h"
+#include "utils/intrusive_list_helpers.h"
+#include "utils/mutex.h"
+#include "utils/rwlock.h"
+
+#include <seastar/core/shared_future.hh>
+#include <seastar/util/defer.hh>
+
+#include <bit>
+
+// Befriended to expose internal state in tests.
+struct test_fixture;
+
+namespace cluster {
+
+template<class Func>
+concept AcceptsUnits = requires(Func f, ssx::semaphore_units units) {
+    f(std::move(units));
+};
+
+class producer_state_manager;
+class producer_state;
+class request;
+
+using producer_ptr = ss::lw_shared_ptr<producer_state>;
+// Note: A shared_promise doesn't guarantee available() to be true
+// right after set_value(), this is an implementation quirk, be
+// mindful of that behavior when using it. We have a test for
+// it in expiring_promise_test
+using result_promise_t = ss::shared_promise<result<kafka_result>>;
+using request_ptr = ss::lw_shared_ptr<request>;
+using seq_t = int32_t;
+
+enum class request_state : uint8_t { NEW = 0, IN_PROGRESS = 1, COMPLETED = 2 };
```

Comment:
lets not use capital names for formatting consistency, we everywhere use non capital letters for enum values

Reply:
fixed.. since new is a keyword, switched it to  initialized